### ISSUE TITLE
Microsoft.OData.ODataException: Nested collection instances are not allowed

### ISSUE THREAD

**Issue by akshukla25** [OData/odata.net#2499]
Seeing the below exception:

### Assemblies affected

OData .Net lib 7.9.3

Looks like validation for ODataCollectionValue is throwing the exception:

Microsoft.OData.Client.DataServiceRequestException: An error occurred while processing this request. ---> Microsoft.OData.ODataException: Nested collection instances are not allowed.
STACK TRACE: 

   at Microsoft.OData.ValidationUtils.ValidateCollectionItem(Object item, Boolean isNullable)
   at Microsoft.OData.JsonLight.ODataJsonLightPropertyAndValueDeserializer.ReadCollectionValue(IEdmCollectionTypeReference collectionValueTypeReference, String payloadTypeName, ODataTypeAnnotation typeAnnotation)
   at Microsoft.OData.JsonLight.ODataJsonLightPropertyAndValueDeserializer.ReadNonEntityValueImplementation(String payloadTypeName, IEdmTypeReference expectedTypeReference, PropertyAndAnnotationCollector propertyAndAnnotationCollector, CollectionWithoutExpectedTypeValidator collectionValidator, Boolean validateNullValue, Boolean isTopLevelPropertyValue, Boolean insideResourceValue, String propertyName, Nullable`1 isDynamicProperty)
   at Microsoft.OData.ODataUriConversionUtils.ConvertFromResourceOrCollectionValue(String value, IEdmModel model, IEdmTypeReference typeReference)
   at Microsoft.OData.ODataUriUtils.ConvertFromUriLiteral(String value, ODataVersion version, IEdmModel model, IEdmTypeReference typeReference)
   at Microsoft.OData.Client.CommonUtil.ParseJsonToPrimitiveValue(String rawValue)
   at Microsoft.OData.Client.Materialization.StructuralValueMaterializationPolicy.MaterializeDynamicProperty(ODataProperty property, Object instance)
   at Microsoft.OData.Client.Materialization.EntryValueMaterializationPolicy.MaterializeResolvedEntry(MaterializerEntry entry, Boolean includeLinks)
   at Microsoft.OData.Client.Materialization.ODataEntityMaterializerInvoker.DirectMaterializePlan(Object materializer, Object entry, Type expectedEntryType)
   at Microsoft.OData.Client.Materialization.ODataEntityMaterializer.ReadImplementation()
   at Microsoft.OData.Client.MaterializeAtom.MoveNextInternal()
   at Microsoft.OData.Client.MaterializeAtom.MoveNext()
   at Microsoft.OData.Client.BaseSaveResult.MaterializeResponse(EntityDescriptor entityDescriptor, ResponseInfo responseInfo, String etag)
   at Microsoft.OData.Client.BaseSaveResult.HandleResponsePost(EntityDescriptor entityDescriptor, String etag)
   at Microsoft.OData.Client.BaseSaveResult.HandleResponsePost(Descriptor descriptor, HeaderCollection contentHeaders)
   at Microsoft.OData.Client.SaveResult.HandleResponse()
 --- End of inner exception stack trace ---
    at Microsoft.OData.Client.SaveResult.HandleResponse()
   at Microsoft.OData.Client.BaseSaveResult.EndRequest()
   at Microsoft.OData.Client.DataServiceContext.SaveChanges(SaveChangesOptions options)
   at Microsoft.Online.Test.DirectoryApi.ProvisioningEngine.RestProvisioningHandler.CreateObject[T](T restObjectValue, IDictionary`2& headers)
 

Looking at this code [https://github.com/OData/odata.net/blob/b37a90e32319636c368a2ce9e5ba2cf39cb65582/src/Microsoft.OData.Core/ValidationUtils.cs#L134](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOData%2Fodata.net%2Fblob%2Fb37a90e32319636c368a2ce9e5ba2cf39cb65582%2Fsrc%2FMicrosoft.OData.Core%2FValidationUtils.cs%23L134&data=05%7C01%7CAkanksha.Shukla%40microsoft.com%7C9f66e399da0d453bf34708da8aad178c%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637974773274884480%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=dLH2GUvwCfiMucpt0WiJWh8%2FrRhMsH9%2BQ3FuTvOL5Sk%3D&reserved=0) but not sure why OData Collection Value is not supported here. Could you please help us understand this and suggest what can we do to fix this?

**Comment by akshukla25** [comment]
Migrated client proxy version 3.0 to client proxy version 4.0 using https://www.nuget.org/packages/Microsoft.OData.Cli which is used for running the extended tests. The client proxy is up to date as the new proxy file is auto generated every time with the build. All the calls failing are scoped to one entity i.e. User. There is nothing unique about this. The entity is not an open type.
Request: Create User -> POST https://directory.msol-test.com/v2/s9h7o2a6e0.msol-test.com/users

**Comment by akshukla25** [comment]
I tried to compare the tests failing vs successful, only difference was that this failure had some properties that were skipped in metadata and proxy file. So, could this be an issue when we have dynamic properties with ODataCollection value?

**Comment by corranrogue9** [comment]
From the [CSDL standard](http://docs.oasis-open.org/odata/odata-csdl-xml/v4.01/odata-csdl-xml-v4.01.html#sec_Type), the type of a collection property must be either a primitive, complex type, or enumeration type. Having a nested collections doesn't really make sense, the nested collection would need to be a property on a complex type. Can you please share your request body, and can you please share the CSDL that your client code is using?

**Comment by akshukla25** [comment]
[RequestBody.txt](https://github.com/OData/odata.net/files/9501807/RequestBody.txt)
[Metadata.txt](https://github.com/OData/odata.net/files/9501810/Metadata.txt)

**Comment by akshukla25** [comment]
We use OData client proxy for running extended tests. Even if populate single property like displayName, all the other properties are sent in the request body. This was a problem for privileged properties because after a privileged property is included in the request body then our service used to throw an exception if the token didn't have access. So, the solution added was to skip adding the privileged properties on the metadata and proxy file so that a request body won't every contain a privileged property. This worked file for OData V3 proxy. But as we migrated to OData V4 properties and updated OData client library, we started seeing this issue. The privileged property like ExternalUserInformation having collection of complex type is written in the response (as it's part of Graph model) and because this does not exist on metadata/proxy, OData client lib populates this as dynamic properties. And I see that the nested collection of complex type on dynamic property is not supported.

<EntityType Name="User" BaseType="Microsoft.DirectoryServices.DirectoryObject" OpenType="true">
<Property Name="externalUserInformation" Type="Microsoft.DirectoryServices.ExternalUserInformation" />
</EntityType>
...
<ComplexType Name="ExternalUserInformation">
        <Property Name="acceptedAs" Type="Edm.String" />
        <Property Name="acceptedOn" Type="Edm.DateTimeOffset" />
        <Property Name="invitedAsMail" Type="Edm.String" />
        <Property Name="invitedOn" Type="Edm.DateTimeOffset" />
        <Property Name="inviteReplyUrls" Type="**Collection(Microsoft.DirectoryServices.KeyValue)"** Nullable="false" />
        <Property Name="inviteResources" Type="Collection(Edm.String)" Nullable="false" />
        <Property Name="signInNames" Type="Collection(Edm.String)" Nullable="false" />
      </ComplexType>

In short, the request body does not contain this property. But on our service side, this property is written in the response (as the token gets access). I think because the metadata/proxy does not contain this property, this property becomes the dynamic property while OData client lib handles the response. And the collection of complex type on dynamic complex type property is throwing the exception.
